// rc-datepicker/lib/Picker.js
'use strict';

exports.__esModule = true;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _utilsDateUtilsJs = require('./utils/DateUtils.js');

var _utilsDateUtilsJs2 = _interopRequireDefault(_utilsDateUtilsJs);

var Picker = _react2['default'].createClass({
  displayName: 'Picker',

  propTypes: {
    date: _react.PropTypes.any.isRequired,
    minDate: _utilsDateUtilsJs2['default'].evaluateDateProp,
    maxDate: _utilsDateUtilsJs2['default'].evaluateDateProp,
    isSelected: _react.PropTypes.bool.isRequired,
    isCurrent: _react.PropTypes.bool.isRequired,
    isEnabled: _react.PropTypes.bool.isRequired,
    isDisabled: _react.PropTypes.bool,
    onSelectDate: _react.PropTypes.func.isRequired,
    mode: _react.PropTypes.string.isRequired
  },

  handleClick: function handleClick(e) {
    e.preventDefault();
    if (this.props.isEnabled) {
      this.props.onSelectDate(this.props.date);
    }
  },

  render: function render() {
    var _classNames;

    var formatMode = undefined;

    switch (this.props.mode) {
      case 'day':
        formatMode = 'D';
        break;

      case 'month':
        formatMode = 'MMM';
        break;

      case 'year':
        formatMode = 'YYYY';
        break;
    }
    var string = this.props.date.format(formatMode);
    var value = string.charAt(0).toUpperCase() + string.slice(1); // first letter always uppercase

    var classes = _classnames2['default']((_classNames = {}, _classNames[this.props.mode] = true, _classNames.current = this.props.isCurrent, _classNames.selected = this.props.isSelected, _classNames.disabled = !this.props.isEnabled, _classNames));

    return _react2['default'].createElement(
      'div',
      { className: 'react-datepicker-picker react-datepicker-button ' + classes, onClick: this.handleClick },
      _react2['default'].createElement(
        'span',
        null,
        value
      )
    );
  }
});

exports['default'] = Picker;
module.exports = exports['default'];
