// rc-datepicker/lib/PickerTop.js
'use strict';

exports.__esModule = true;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

exports['default'] = _react2['default'].createClass({
  displayName: 'PickerTop',

  propTypes: {
    fixed: _react.PropTypes.bool,
    handleClick: _react.PropTypes.func,
    nextDate: _react.PropTypes.func,
    previousDate: _react.PropTypes.func,
    value: _react.PropTypes.oneOfType([_react.PropTypes.string, _react.PropTypes.number]).isRequired,
    valueClassName: _react.PropTypes.string,
    weekDays: _react.PropTypes.element
  },

  getDefaultProps: function getDefaultProps() {
    return {
      handleClick: function handleClick() {},
      nextDate: function nextDate() {},
      previousDate: function previousDate() {}
    };
  },

  render: function render() {
    return _react2['default'].createElement(
      'div',
      { className: 'react-datepicker-top' },
      _react2['default'].createElement(
        'div',
        { className: 'display' },
        _react2['default'].createElement(
          'div',
          { className: 'react-datepicker-button button-left',
            onClick: this.props.previousDate
          },
          '<'
        ),
        _react2['default'].createElement(
          'div',
          { className: _classnames2['default']('react-datepicker-button button-label', { fixed: this.props.fixed }),
            onClick: this.props.handleClick
          },
          _react2['default'].createElement(
            'strong',
            { className: this.props.valueClassName },
            this.props.value
          )
        ),
        _react2['default'].createElement(
          'div',
          { className: 'react-datepicker-button button-right',
            onClick: this.props.nextDate
          },
          '>'
        )
      ),
      this.props.weekDays
    );
  }
});
module.exports = exports['default'];
