// rc-datepicker/lib/daypicker/DayPicker.js
'use strict';

exports.__esModule = true;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _utilsDateUtilsJs = require('../utils/DateUtils.js');

var _utilsDateUtilsJs2 = _interopRequireDefault(_utilsDateUtilsJs);

var _DayPickerTop = require('./DayPickerTop');

var _DayPickerTop2 = _interopRequireDefault(_DayPickerTop);

var _DayPickerBody = require('./DayPickerBody');

var _DayPickerBody2 = _interopRequireDefault(_DayPickerBody);

var DayPicker = _react2['default'].createClass({
  displayName: 'DayPicker',

  propTypes: {
    changeMonth: _react.PropTypes.func.isRequired,
    visibleDate: _react.PropTypes.any.isRequired,
    date: _utilsDateUtilsJs2['default'].evaluateDateProp,
    minDate: _utilsDateUtilsJs2['default'].evaluateDateProp,
    maxDate: _utilsDateUtilsJs2['default'].evaluateDateProp,
    onSelectDate: _react.PropTypes.func.isRequired,
    onChangeMode: _react.PropTypes.func.isRequired,
    mode: _react.PropTypes.string.isRequired,
    fixedMode: _react2['default'].PropTypes.bool
  },

  _onSelectDate: function _onSelectDate(date) {
    this.props.onSelectDate(date);
  },

  render: function render() {
    return _react2['default'].createElement(
      'div',
      { className: 'react-datepicker-container day' },
      _react2['default'].createElement(_DayPickerTop2['default'], {
        changeMonth: this.props.changeMonth,
        visibleDate: this.props.visibleDate,
        onChangeMode: this.props.onChangeMode,
        fixedMode: this.props.fixedMode
      }),
      _react2['default'].createElement(_DayPickerBody2['default'], {
        visibleDate: this.props.visibleDate,
        date: this.props.date,
        minDate: this.props.minDate,
        maxDate: this.props.maxDate,
        onSelectDate: this._onSelectDate,
        mode: this.props.mode
      })
    );
  }
});

exports['default'] = DayPicker;
module.exports = exports['default'];
