// rc-datepicker/lib/yearpicker/YearPickerTop.js
'use strict';

exports.__esModule = true;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _PickerTop = require('../PickerTop');

var _PickerTop2 = _interopRequireDefault(_PickerTop);

var _lodashPartial = require('lodash/partial');

var _lodashPartial2 = _interopRequireDefault(_lodashPartial);

var YearPickerTop = _react2['default'].createClass({
  displayName: 'YearPickerTop',

  propTypes: {
    visibleDate: _react.PropTypes.any.isRequired,
    changeYear: _react.PropTypes.func.isRequired,
    textClassNames: _react.PropTypes.string
  },

  render: function render() {
    var year = this.props.visibleDate.year();
    var startDecadeYear = parseInt(year / 10, 10) * 10;
    var endDecadeYear = startDecadeYear + 9;
    return _react2['default'].createElement(_PickerTop2['default'], {
      fixed: true,
      nextDate: _lodashPartial2['default'](this.props.changeYear, year + 10),
      previousDate: _lodashPartial2['default'](this.props.changeYear, year - 10),
      value: startDecadeYear + '-' + endDecadeYear,
      valueClassName: this.props.textClassNames
    });
  }
});

exports['default'] = YearPickerTop;
module.exports = exports['default'];
