// rc-datepicker/lib/yearpicker/YearPicker.js
'use strict';

exports.__esModule = true;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _utilsDateUtilsJs = require('../utils/DateUtils.js');

var _utilsDateUtilsJs2 = _interopRequireDefault(_utilsDateUtilsJs);

var _YearPickerTop = require('./YearPickerTop');

var _YearPickerTop2 = _interopRequireDefault(_YearPickerTop);

var _YearPickerBody = require('./YearPickerBody');

var _YearPickerBody2 = _interopRequireDefault(_YearPickerBody);

var YearPicker = _react2['default'].createClass({
  displayName: 'YearPicker',

  propTypes: {
    changeYear: _react.PropTypes.func.isRequired,
    visibleDate: _react.PropTypes.any.isRequired,
    date: _utilsDateUtilsJs2['default'].evaluateDateProp,
    minDate: _utilsDateUtilsJs2['default'].evaluateDateProp,
    maxDate: _utilsDateUtilsJs2['default'].evaluateDateProp,
    onSelectDate: _react.PropTypes.func.isRequired,
    onChangeMode: _react.PropTypes.func.isRequired,
    onChangeVisibleDate: _react.PropTypes.func.isRequired,
    mode: _react.PropTypes.string.isRequired,
    fixedMode: _react.PropTypes.bool
  },

  _onSelectDate: function _onSelectDate(date) {
    if (this.props.fixedMode) {
      this.props.onSelectDate(date);
    } else {
      this.props.onChangeVisibleDate(date);
      this.props.onChangeMode('month');
    }
  },

  render: function render() {
    return _react2['default'].createElement(
      'div',
      { className: 'react-datepicker-container year' },
      _react2['default'].createElement(_YearPickerTop2['default'], {
        changeYear: this.props.changeYear,
        visibleDate: this.props.visibleDate
      }),
      _react2['default'].createElement(_YearPickerBody2['default'], {
        visibleDate: this.props.visibleDate,
        date: this.props.date,
        minDate: this.props.minDate,
        maxDate: this.props.maxDate,
        onSelectDate: this._onSelectDate,
        mode: this.props.mode
      })
    );
  }
});

exports['default'] = YearPicker;
module.exports = exports['default'];
