// rc-datepicker/lib/yearpicker/YearPickerBody.js
'use strict';

exports.__esModule = true;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _moment = require('moment');

var _moment2 = _interopRequireDefault(_moment);

var _InvalidDate = require('../InvalidDate');

var _InvalidDate2 = _interopRequireDefault(_InvalidDate);

var _Picker = require('../Picker');

var _Picker2 = _interopRequireDefault(_Picker);

var _Row = require('../Row');

var _Row2 = _interopRequireDefault(_Row);

var _utilsDateUtils = require('../utils/DateUtils');

var _utilsDateUtils2 = _interopRequireDefault(_utilsDateUtils);

var _lodashRange = require('lodash/range');

var _lodashRange2 = _interopRequireDefault(_lodashRange);

var YearPickerBody = _react2['default'].createClass({
  displayName: 'YearPickerBody',

  propTypes: {
    visibleDate: _react.PropTypes.any.isRequired,
    date: _utilsDateUtils2['default'].evaluateDateProp,
    minDate: _utilsDateUtils2['default'].evaluateDateProp,
    maxDate: _utilsDateUtils2['default'].evaluateDateProp,
    onSelectDate: _react.PropTypes.func.isRequired,
    mode: _react.PropTypes.string.isRequired
  },

  render: function render() {
    var _this = this;

    if (!this.props.visibleDate.isValid()) {
      return _react2['default'].createElement(_InvalidDate2['default'], { invalidDate: this.props.visibleDate.format() });
    }
    var year = this.props.visibleDate.year();
    var selectedYear = this.props.date ? this.props.date.year() : -1;

    var visibleYears = _utilsDateUtils2['default'].getVisibleYears(year);
    var years = visibleYears.years.map(function (_year, index) {
      var date = _moment2['default']([_year, 0, 1]);
      var isCurrent = index >= visibleYears.startCurrent && index <= visibleYears.endCurrent;
      return _react2['default'].createElement(_Picker2['default'], {
        date: date,
        isSelected: selectedYear === _year,
        isCurrent: isCurrent,
        isEnabled: _utilsDateUtils2['default'].isInsideTheEnabledArea(date, _this.props.mode, _this.props.minDate, _this.props.maxDate),
        onSelectDate: _this.props.onSelectDate,
        mode: _this.props.mode,
        key: index
      });
    });
    var nColumns = 4;
    var nRows = 3;
    var rows = _lodashRange2['default'](nRows).map(function (index) {
      return _react2['default'].createElement(_Row2['default'], { pickers: years.slice(nColumns * index, nColumns * (index + 1)), mode: _this.props.mode, key: index });
    });

    return _react2['default'].createElement(
      'div',
      { className: 'react-datepicker-body' },
      rows
    );
  }
});

exports['default'] = YearPickerBody;
module.exports = exports['default'];
