// rc-datepicker/lib/utils/DateUtils.js
'use strict';

exports.__esModule = true;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _moment = require('moment');

var _moment2 = _interopRequireDefault(_moment);

var _lodashRange = require('lodash/range');

var _lodashRange2 = _interopRequireDefault(_lodashRange);

var daysInMonthCount = function daysInMonthCount(month, year) {
  return _moment2['default']([year, month]).endOf('month').date();
};

var getArrayByBoundary = function getArrayByBoundary(start, end) {
  return _lodashRange2['default'](end - start).map(function (i) {
    return i + start;
  });
};

var getWeekdaysMin = function getWeekdaysMin() {
  var offset = _moment2['default']().localeData().firstDayOfWeek();
  var weekdaysMin = _moment2['default'].weekdaysMin();

  _lodashRange2['default'](offset).forEach(function () {
    var firstDay = weekdaysMin.shift();
    weekdaysMin.push(firstDay);
  });
  return weekdaysMin;
};

var getVisibleDays = function getVisibleDays(month, year) {
  var offset = _moment2['default']([year, month]).startOf('month').weekday();
  var previousMonth = month === 0 ? 11 : month - 1;
  var previousYear = month === 0 ? year - 1 : year;

  var currentMonthLength = daysInMonthCount(month, year) + 1;
  var previousMonthLength = daysInMonthCount(previousMonth, previousYear) + 1; // We need the last number too

  var previous = getArrayByBoundary(previousMonthLength - offset, previousMonthLength);
  var current = getArrayByBoundary(1, currentMonthLength);
  var following = getArrayByBoundary(1, 43 - previous.length - current.length);
  return {
    startCurrent: previous.length,
    endCurrent: previous.length + current.length - 1,
    days: previous.concat(current).concat(following)
  };
};

var getVisibleYears = function getVisibleYears(year) {
  var startDecadeYear = parseInt(year / 10, 10) * 10;
  var endDecadeYear = startDecadeYear + 9;
  var previous = [startDecadeYear - 1];
  var current = getArrayByBoundary(startDecadeYear, endDecadeYear + 1);
  var following = [endDecadeYear + 1];
  return {
    startCurrent: previous.length,
    endCurrent: previous.length + current.length - 1,
    years: previous.concat(current).concat(following)
  };
};

var evaluateDateProp = function evaluateDateProp(props, propName, componentName) {
  var dateProp = props[propName];
  if (dateProp && typeof dateProp !== 'string' && !(dateProp instanceof Date) && !_moment2['default'].isMoment(dateProp)) {
    return new Error(propName + ' validation failed in ' + componentName);
  }
};

var isInsideTheEnabledArea = function isInsideTheEnabledArea(date, mode, minDate, maxDate) {
  if (!minDate && !maxDate) {
    return true;
  }

  var minDateMoment = typeof minDate === 'string' ? _moment2['default'](minDate, _moment2['default'].ISO_8601, true) : _moment2['default'](minDate);
  var maxDateMoment = typeof maxDate === 'string' ? _moment2['default'](maxDate, _moment2['default'].ISO_8601, true) : _moment2['default'](maxDate);

  var format = undefined;
  switch (mode) {
    case 'day':
      format = 'YYYY/MM/DD';
      break;

    case 'month':
      format = 'YYYY/MM';
      break;

    case 'year':
      format = 'YYYY';
      break;
  }

  return (!minDate || date.format(format) >= minDateMoment.format(format)) && (!maxDate || date.format(format) <= maxDateMoment.format(format));
};

exports['default'] = {
  daysInMonthCount: daysInMonthCount,
  getVisibleDays: getVisibleDays,
  getVisibleYears: getVisibleYears,
  getWeekdaysMin: getWeekdaysMin,
  evaluateDateProp: evaluateDateProp,
  isInsideTheEnabledArea: isInsideTheEnabledArea,
  getArrayByBoundary: getArrayByBoundary
};
module.exports = exports['default'];
